### PR TITLE
Replace db::Label with static method

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- The `db::Label label` in tags is replaced with `std::string name() noexcept`. This allows more flexibility for naming tags. This is useful for retrieving items from the DataBox using the string for observing.
- Add IWYU make targets

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- Code has documentation and unit tests
- Private member variables have a trailing underscore
- Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- Header order:
  1. hpp corresponding to cpp (only in cpp files) or `tests/Unit/TestingFramework.hpp` (only in tests)
  2. Blank line (only in cpp files and tests)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- File lists in CMake are alphabetical
- Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- Mark objects `const` whenever possible
- Almost always `auto`, except with expression templates, i.e. `DataVector`
- All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- Prefix commits addressing PR requests with `fixup`. These commits are flagged
  by Travis so we do not accidentally merge them.
- Include what you use, prefer forward declarations in hpp files.
- Explicitly make numbers floating point, e.g. `2.` or `2.0` over `2`
- Use `Tensor`'s non-member get if the indices are constant expressions, e.g.
  `get<1>(tensor)`

### CODE REVIEW

File: src/ErrorHandling/StaticAssert.hpp [hunk 1]
```diff
@@ -13,5 +13,6 @@
 #ifdef SPECTRE_DEBUG
 #define DEBUG_STATIC_ASSERT(...) static_assert(__VA_ARGS__)
 #else   // ifdef  SPECTRE_DEBUG
-#define DEBUG_STATIC_ASSERT(...)
+// GCC warns of an unnecessary semi-colon without the static_assert(true)
+#define DEBUG_STATIC_ASSERT(...) static_assert(true, "")
```
Comment (kidder) [inline src/ErrorHandling/StaticAssert.hpp]: drop and be smart

File: src/DataStructures/DataBox/DataBox.hpp [hunk 2]
```diff
@@ -324,6 +325,32 @@ struct create_dependency_graph<TagList, Caller, Callee, List,
...
+struct tag_has_name<
+    Tag, Requires<cpp17::is_same_v<decltype(Tag::name()), std::string>>> {};
+
+template <typename Tag, typename = std::nullptr_t>
+struct check_simple_or_compute_tag {
+  static_assert(cpp17::is_same_v<Tag, const void* const*>,
+                "All tags added to a DataBox must derive off of db::SimpleTag "
+                "or db::ComputeTag, you cannot add a base tag itself. See the "
+                "first template parameter of "
+                "db::DataBox_detail::check_simple_or_compute_tag to see what "
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: remove what

File: src/DataStructures/DataBox/DataBox.hpp [hunk 3]
```diff
@@ -419,7 +438,18 @@ class DataBox<tmpl::list<Tags...>>
     }
     return *this;
   }
+#ifdef SPECTRE_DEBUG
+  // Destructor is used for triggering assertions
+  ~DataBox() {
```
Comment (kidder) [inline src/DataStructures/DataBox/DataBox.hpp]: noexcept and use EXPAND...

File: src/DataStructures/Tensor/Structure.hpp [hunk 4]
```diff
@@ -13,7 +13,7 @@
 #include "DataStructures/Tensor/Metafunctions.hpp"
 #include "DataStructures/Tensor/Symmetry.hpp"
 #include "ErrorHandling/Assert.hpp"
-#include "Utilities/Array.hpp"
+#include "Utilities/Array.hpp"  // IWYU pragma: export
```
Comment (kidder) [inline src/DataStructures/Tensor/Structure.hpp:16]: remove this

File: tests/Unit/DataStructures/Test_DataBox.cpp [hunk 5]
```diff
@@ -144,14 +146,10 @@ static_assert(
...
               "Failed testing no-op create_from");
-
-static_assert(db::DataBox_detail::tag_has_label_v<test_databox_tags::Tag0>,
-              "Failed testing db::tag_has_label");
-static_assert(db::DataBox_detail::tag_has_label_v<test_databox_tags::TagTensor>,
-              "Failed testing db::tag_has_label");
 }  // namespace
 
 SPECTRE_TEST_CASE("Unit.DataStructures.DataBox", "[Unit][DataStructures]") {
+  // db::DataBox_detail::check_has_tag_name_helper<test_databox_tags::Tag0>();
```
Comment (kidder) [inline tests/Unit/DataStructures/Test_DataBox.cpp]: remove

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — COMMENTED** [review]
squash

**Review by kidder — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**